#N/A: automated modifier statement combining

ntseng opened the pull request:
Summary
-------
- added automated modifier statement combining and refactored move effects to use it
- `addModifier()`, `removeModifier()` and `enterStance()` now return `ModifierReceipt[]` instead of their previous returns
- moved Progress logic into Elkemist's actions

Created a new class "ModifierReceipt" to standardize modifier addition or removal attempts, while including enough data for moves with logic based on modifier addition or removal to function and report unsuccessful modifier change attempts (ie Retain and Oblivious). This standardized shape also allowed for the creation of functions to handle them in bulk: `combineModifierReceipts()` for minimizing the number of lines mentioning a specific combatant, and `generateModifierResultLines()` for converting receipts into user-readable text.

Future work should include creating sibling receipt classes for damage, healing, stagger, and protection, as there is still some need for custom result line generation when combining those changes to combatants.

Local Tests Performed
---------------------
- [x] bot still turns on (no BuildErrors or circular dependencies)
- [x] Successful run through the Debug Dungeon

ntseng committed:
automated modifier statement combining

doubledh commented on source/util/mathUtil.js:
```diff
@@ -9,7 +9,24 @@ function anyDieSucceeds(successChance, extraDice) {
+		return false;
+	}
+
+	for (const element of firstSet) {
```
could use [Set.prototype.isSubsetOf()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/isSubsetOf)


or


[Set.prototype.isSupersetOf()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/isSupersetOf)

or check the length of
[Set.prototype.symmetricDifference()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set/symmetricDifference)

though it'd likely be the same execution/runtime 🤷 so I'd take preference  of your already-done work

ntseng replied:
I originally wanted to use `isSubsetOf()`, but was scared off thinking it might not be in the node version I was running because VS Code wasn't auto-completing it. (It does work in the node REPL though)

doubledh commented on source/util/combatantUtil.js:
```diff
@@ -207,26 +205,13 @@ function addModifier(combatants, { name: modifier, stacks: pendingStacks, force
-			} else {
-				affectedCombatants.push(combatant.name);
-			}
+			receipts.push(new ModifierReceipt(combatant.name, "add", [getApplicationEmojiMarkdown(modifier)], []));
```
is it normal that it's mechanically impossible to directly remove obviousness (in the hopes of removing multiple stacks)

ntseng replied:
`removeModifier(combatants, {name: "Oblivious", stacks: "all", force: true})` would remove all Oblivious in one call. `force: true` is usually reserved for game rule application, but removing all stacks of Oblivious would likely be that?

doubledh commented on source/enemies/mechaqueenmech.js:
```diff
@@ -1,6 +1,6 @@
 const { EnemyTemplate } = require("../classes/index.js");
 const { selectRandomFoe, selectNone, selectAllFoes, selectRandomOtherAlly, selectAllAllies } = require("../shared/actionComponents.js");
-const { addModifier, dealDamage, changeStagger, addProtection } = require("../util/combatantUtil.js");
+const { addModifier, dealDamage, changeStagger, addProtection, combineModifierReceipts } = require("../util/combatantUtil.js");
```
import generateModifierResultLines?

ntseng replied:
Good catch!

doubledh reviewed (changes requested):
LGTM except for one missing import on queen
- not entirely sure the differences between `receipts.push(...removeModifier`, and unshift, and concat
- not "ideal" on combining both buffs/targets but will do its job for centralization, and any more ideal is probably towards hair-splitting and diminishing returns

ntseng committed:
add missing import

doubledh reviewed (approved):
yayfix

ntseng committed:
playtest fixes
- fix unneeded initial receipt offset
- bold result lines for better readability
- fix missing import
- functionize lives result line generation
